### PR TITLE
teerex refactor 2

### LINKED ISSUE #186: [teerex] unify dispatchables for enclave registration
storage
image

image

image

dispatchables
image

### PR DESCRIPTION
**Pull request by brenzi**
closes #186 #183 #152 

breaking changes:
* ShardStatus will replace WorkerForShard logic
* remove skip-ias
* introduce MultiEnclave

non-breaking changes:
* no automatic garbage collection OnTimestamp anymore. lazy dispatchable introduced

### CODE REVIEW

**Comment by brenzi** [comment]
I just realize that refactoring the sidechain pallet will likely require to solve #183 because we currently only let the first ever registered validateer finalize and the move to a storage map no longer let's us derive an order

**Comment by brenzi** [comment]
breaking change: `ShardStatus` will replace `WorkerForShard` logic

File: primitives/teerex/Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,7 @@ serde = { version = "1.0.13", default-features = false }
 # substrate dependencies
 sp-core = { default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.42" }
 sp-io = { default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.42" }
+sp-runtime = { default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.42" }
```
Comment (clangenb) [inline primitives/teerex/Cargo.toml:19]: std enablement missing

File: primitives/teerex/src/lib.rs [hunk 2]
```diff
@@ -92,6 +93,94 @@ pub enum SgxStatus {
...
+pub enum AnySigner {
+	Opaque(OpaqueSigner),
+	Known(MultiSigner),
+}
+
+impl From<MultiSigner> for AnySigner {
+	fn from(signer: MultiSigner) -> Self {
+		AnySigner::Known(signer)
+	}
+}
```
Comment (clangenb) [inline primitives/teerex/src/lib.rs:109]: We could remove these manual implementations by importing `derive_more` and then derive `From`

File: teerex/src/lib.rs [hunk 3]
```diff
@@ -162,70 +169,143 @@ pub mod pallet {
...
+							T::AccountId::decode(&mut report.report_data.lower32().as_ref()),
+						<Error<T>>::SenderIsNotAttestedEnclave
+					);
+
+					// TODO: activate state checks as soon as we've fixed our setup #83
+					// ensure!((report.status == SgxStatus::Ok) | (report.status == SgxStatus::ConfigurationNeeded),
+					//     "RA status is insufficient");
+					// log::info!("teerex: status is acceptable");
+
+					Self::ensure_timestamp_within_24_hours(report.timestamp)?;
```
Comment (clangenb) [inline teerex/src/lib.rs:213]: Intuitively, it makes more sense if this is directly after verifying the report.

File: teerex/src/lib.rs [hunk 4]
```diff
@@ -162,70 +169,143 @@ pub mod pallet {
...
+			ensure!(
+				worker_url.clone().unwrap_or_default().len() <= MAX_URL_LEN,
+				<Error<T>>::EnclaveUrlTooLong
+			);
 			log::info!("teerex: parameter length ok");
 
-			#[cfg(not(feature = "skip-ias-check"))]
-			let enclave = Self::verify_report(&sender, ra_report)?.with_url(worker_url.clone());
+			let enclave = match attestation_method {
+				SgxAttestationMethod::Ias => {
```
Comment (clangenb) [inline teerex/src/lib.rs:188]: This whole function will be much more readable if we have something like:

```rust
let enclave = match attestation_method {
   SgxAttestationMethod::Ias => verify_ias_attestation(...)?,
   SgxAttestationMethod::Dcap { proxied } => verify_dcap_quote(...)?,
   SgxAttestationMethod::Skipped { proxied } => handle_skipped_attestation(...)?
}


```

File: teerex/src/lib.rs [hunk 5]
```diff
@@ -162,70 +169,143 @@ pub mod pallet {
...
+			log::info!("teerex: called into runtime call unregister_sovereign_enclave()");
 			let sender = ensure_signed(origin)?;
-
-			Self::remove_enclave(&sender)?;
-			Self::deposit_event(Event::RemovedEnclave(sender));
+			let enclave = Self::sovereign_enclaves(&enclave_signer)
+				.ok_or(<Error<T>>::EnclaveIsNotRegistered)?;
+			let now = <timestamp::Module<T>>::get();
+			let oldest_acceptable_attestation_time =
+				now.saturating_sub(T::MaxSilenceTime::get()).saturated_into::<u64>();
```
Comment (clangenb) [inline teerex/src/lib.rs:304]: I think we should rename this to `T::AttestationTimeout` or something like this.

File: teerex/src/lib.rs [hunk 6]
```diff
@@ -474,7 +500,7 @@ pub mod pallet {
 		/// The worker url is too long.
 		EnclaveUrlTooLong,
 		/// The Remote Attestation report is too long.
-		RaReportTooLong,
+		RaProofTooLong,
```
Comment (clangenb) [inline teerex/src/lib.rs:503]: The doc should be adjusted as well

File: teerex/src/lib.rs [hunk 7]
```diff
@@ -717,11 +597,45 @@ impl<T: Config> Pallet<T> {
...
+
+	pub fn most_recent_shard_update(
+		shard: &ShardIdentifier,
+	) -> Option<ShardSignerStatus<T::AccountId, T::BlockNumber>> {
+		if let Some(mut statuses) = <ShardStatus<T>>::get(shard) {
+			statuses.sort_by_key(|a| a.last_activity);
+			statuses.last().cloned()
+		} else {
+			None
+		}
```
Comment (clangenb) [inline teerex/src/lib.rs:638]:
```suggestion
		<ShardStatus<T>>::get(shard).map(|statuses| {
			statuses.sort_by_key(|a| a.last_activity);
			statuses.last().cloned()
		})
```

File: teerex/src/lib.rs [hunk 8]
```diff
@@ -717,11 +597,45 @@ impl<T: Config> Pallet<T> {
...
+		let signer_statuses = if let Some(mut status_vec) = <ShardStatus<T>>::get(shard) {
+			if let Some(index) = status_vec.iter().position(|i| i.signer == *enclave_signer) {
+				status_vec[index] = fresh_status;
+			} else {
+				status_vec.push(fresh_status)
+			}
+			status_vec
+		} else {
+			vec![fresh_status]
+		};
```
Comment (clangenb) [inline teerex/src/lib.rs:625]:
```suggestion
		let signer_statuses = get_or_init(...);
		
// more efficient, and maybe even more readable. :)
pub fn get_or_init(...) {
  <ShardStatus<T>>::get(shard)
  .map(|status_vec| {
      status_vec.iter().position(|i| i.signer == *enclave_signer)
        .map_or_else(|| status_vec.push(free_status), |i| status_vec[i] = fresh_status);
      
      status_vec
  }).unwrap_or_else(|| vec![fresh_status])
}
```

File: teerex/src/lib.rs [hunk 9]
```diff
@@ -717,11 +597,45 @@ impl<T: Config> Pallet<T> {
...
+	pub fn poke_shard(
+		shard: ShardIdentifier,
+		enclave_signer: &T::AccountId,
+	) -> Result<ShardSignerStatuses<T>, DispatchErrorWithPostInfo> {
+		let enclave = Self::sovereign_enclaves(enclave_signer.clone())
+			.ok_or(<Error<T>>::EnclaveIsNotRegistered)?;
+
+		let current_block_number = <frame_system::Pallet<T>>::block_number();
+
+		let fresh_status = ShardSignerStatus {
```
Comment (clangenb) [inline teerex/src/lib.rs:610]: `new_status` seems like a more common terminology to me.

File: teerex/src/lib.rs [hunk 10]
```diff
@@ -717,11 +597,45 @@ impl<T: Config> Pallet<T> {
 			Err(<Error<T>>::RemoteAttestationTooOld.into())
 		}
 	}
-}
 
-impl<T: Config> OnTimestampSet<T::Moment> for Pallet<T> {
-	fn on_timestamp_set(moment: T::Moment) {
-		Self::unregister_silent_workers(moment)
+	pub fn poke_shard(
```
Comment (clangenb) [inline teerex/src/lib.rs:601]: From the outside it is a bit hard to guess, what is `poke`, I think the more commonly used term `touch` reflects better, what this function represents.

**Review by clangenb — COMMENTED** [review]
Reviewed everything except for the tests. Looks good so far!

File: teerex/src/lib.rs (hunk 10 above)
Comment (brenzi) [reply teerex/src/lib.rs:601]: Thanks for the intermediate review. I'll iterate and guess its better to wait with the next phase of the review

File: teerex/src/lib.rs (hunk 4 above)
Comment (brenzi) [reply teerex/src/lib.rs:188]: I think readability is debatable. I don't like 3 line dispatchables that do nothing else than calling a fn within the same file....and then dispatch an event. if it's only a few lines I prefer reading them in-place.
Testability would be a stronger argument IMO, but as all unit tests go to the dispatchable level directly, I saw no reason to modularize

File: teerex/src/lib.rs (hunk 5 above)
Comment (brenzi) [reply teerex/src/lib.rs:304]: max silence time also renews upon activity. not only attestation. We could use MaxPassiveTime?

File: primitives/teerex/src/lib.rs (hunk 2 above)
Comment (brenzi) [reply primitives/teerex/src/lib.rs]: much wow! love it!

File: teerex/src/lib.rs (hunk 8 above)
Comment (brenzi) [reply teerex/src/lib.rs]: borrow checker not happy with this. but I adapted pragmatically ;-)

File: teerex/src/lib.rs (hunk 5 above)
Comment (brenzi) [reply teerex/src/lib.rs:306]: not much better IMO

Comment (clangenb) [reply teerex/src/lib.rs:306]: I also thought that this is the design, but when I looked at the code, I could only find that it is renewed upon a new attestation. But in this case, I guess I was wrong, and then it is ok as it is

File: teerex/src/lib.rs (hunk 4 above)
Comment (clangenb) [reply teerex/src/lib.rs:188]: Hmm, I disagree here to be honest, when going down, it needs a lot of focus to even remember, whether you are still in a match branch or not, but I won't argue.

File: teerex/src/lib.rs [hunk 11]
```diff
@@ -162,70 +169,147 @@ pub mod pallet {
...
-			log::info!("teerex: called into runtime call register_ias_enclave()");
+			log::info!("teerex: called into runtime call register_sgx_enclave()");
 			let sender = ensure_signed(origin)?;
-			ensure!(ra_report.len() <= MAX_RA_REPORT_LEN, <Error<T>>::RaReportTooLong);
-			ensure!(worker_url.len() <= MAX_URL_LEN, <Error<T>>::EnclaveUrlTooLong);
+			ensure!(proof.len() <= SGX_RA_PROOF_MAX_LEN, <Error<T>>::RaProofTooLong);
+			ensure!(
+				worker_url.clone().unwrap_or_default().len() <= MAX_URL_LEN,
+				<Error<T>>::EnclaveUrlTooLong
+			);
```
Comment (clangenb) [inline teerex/src/lib.rs:184]: Unnecessary clone can be removed.

Maybe `worker_url.as_ref()` needed in the `if let` statement.
```suggestion
if let Some(url) = worker_url {
	ensure!(url.len() <= MAX_URL_LEN, <Error<T>>::EnclaveUrlTooLong);
}
```

**Review by clangenb — APPROVED** [review]
Alright, looks good to me except for one detail, but I shall approve already. :)

**Pull request merged**